selfdrive/car/nissan/carstate.py

```diff
@@ -0,0 +1,143 @@
+
+def get_powertrain_can_parser(CP):
+  # this function generates lists for signal, messages and initial values
+  signals = [
+    # sig_name, sig_address, default
+    ("FL", "WheelspeedFront", 0),
+    ("FR", "WheelspeedFront", 0),
+    ("RL", "WheelspeedRear", 0),
+    ("RR", "WheelspeedRear", 0),
+    ("DOOR_OPEN_FR", "DoorsLights", 1),
+    ("DOOR_OPEN_FL", "DoorsLights", 1),
+    ("DOOR_OPEN_RR", "DoorsLights", 1),
+    ("DOOR_OPEN_RL", "DoorsLights", 1),
+    ("USER_BRAKE_PRESSED", "DoorsLights", 1),
+    ("BRAKE_LIGHT", "DoorsLights", 1),
+    ("DriverTorque", "Steering", 0),
+    ("DriverTouchingWheel", "STEER_TORQUE", 0),
+    ("ThrottlePedal", "Throttle", 0),
+    ("Steering_Angle", "SteeringWheel", 0),
+    ("RIGHT_BLINKER", "Lights", 0),
+    ("LEFT_BLINKER", "Lights", 0),
+    ("PROPILOT_BUTTON", "CruiseThrottle", 0),
+    ("CANCEL_BUTTON", "CruiseThrottle", 0),
+    ("GAS_PEDAL_INVERTED", "CruiseThrottle", 0),
+    ("unsure2", "CruiseThrottle", 0),
+    ("SET_BUTTON", "CruiseThrottle", 0),
+    ("RES_BUTTON", "CruiseThrottle", 0),
+    ("FOLLOW_DISTANCE_BUTTON", "CruiseThrottle", 0),
+    ("NO_BUTTON_PRESSED", "CruiseThrottle", 0),
+    ("GAS_PEDAL", "CruiseThrottle", 0),
+    ("unsure3", "CruiseThrottle", 0),
+    ("unsure", "CruiseThrottle", 0),
+  ]
+
+  checks = [
+    # sig_address, frequency
+    ("WheelspeedRear", 50),
+    ("WheelspeedFront", 50),
+    ("DoorsLights", 10),
+  ]
+
+  return CANParser(DBC[CP.carFingerprint]['pt'], signals, checks, 0)
+
+
+def get_adas_can_parser(CP):
+  # this function generates lists for signal, messages and initial values
+  signals = [
+    # sig_name, sig_address, default
+    ("Des_Angle", "LKAS", 0),
+    ("SET_0x80_2", "LKAS", 0),
+    ("NEW_SIGNAL_4", "LKAS", 0),
+    ("SET_X80", "LKAS", 0),
+    ("Counter", "LKAS", 0),
+    ("LKA_Active", "LKAS", 0),
+  ]
+
+  checks = [
+    # sig_address, frequency
+  ]
+
+  return CANParser(DBC[CP.carFingerprint]['pt'], signals, checks, 2)
+
+def get_camera_can_parser(CP):
+  signals = [
+    ("CRUISE_ON", "ProPilot", 0),
+    ("CRUISE_ACTIVATED", "ProPilot", 0),
+    ("STEER_STATUS", "ProPilot", 0),
+  ]
+
+  checks = [
+  ]
+
+  return CANParser(DBC[CP.carFingerprint]['pt'], signals, checks, 1)
+
+class CarState(object):
+  def __init__(self, CP):
+    # initialize can parser
+    self.CP = CP
+
+    self.car_fingerprint = CP.carFingerprint
+    self.left_blinker_on = False
+    self.prev_left_blinker_on = False
+    self.right_blinker_on = False
+    self.prev_right_blinker_on = False
+    self.steer_torque_driver = 0
+    self.steer_not_allowed = False
+    self.main_on = False
+
+    # vEgo kalman filter
+    dt = 0.01
+    self.v_ego_kf = KF1D(x0=[[0.], [0.]],
+                         A=[[1., dt], [0., 1.]],
+                         C=[1., 0.],
+                         K=[[0.12287673], [0.29666309]])
+    self.v_ego = 0.
+
+  def update(self, cp, cp_adas, cp_cam):
+
+    self.pedal_gas = cp.vl["Throttle"]["ThrottlePedal"]
```

Comment:
**Recently made improvement in car abstraction (see other cars)**
you should return capnp struct directly